File: practices/chapter12_recursive/python/ch_12_hanoi.py
def hanoi(n, source, aux, target):
  if n == 1:
    # 結束移動
    print('Move {} from {} to {}'.format(n, source, target))
  else:
    # 從來源移到輔助
    hanoi(n - 1, source, target, aux)
    print('Move {} from {} to {}'.format(n, source, target))
    # 從輔助移到目標
    hanoi(n - 1, aux, source, target)

def main():
  hanoi(3, 'A', 'B', 'C')

if __name__ == "__main__":
    main()

File: practices/chapter12_recursive/js/ch_12_hanoi.js
function hanoi(n, source, aux, target) {
  if(n === 1) {
    console.log(`Move ${n} from ${source} to ${target}`)
  } else {
    // 把最小的放到輔助，就等同少一個n的解法(因為可以將輔助視為沒東西)，去解n - 1的河內塔
    hanoi(n - 1, source, target, aux)
    console.log(`Move ${n} from ${source} to ${target}`)
    // 最後再把輔助的放到目標
    hanoi(n - 1, aux, source, target)
  }
}

function main() {
  hanoi(4, 'A', 'B', 'C')
}

main()